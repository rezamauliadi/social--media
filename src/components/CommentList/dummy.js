export default [
  {
    postId: 1,
    id: 1,
    name: "id labore",
    email: "Eliseo@gardner.biz",
    body: "laudantium "
  },
  {
    postId: 1,
    id: 2,
    name: "quo vero ",
    email: "Jayne_Kuhic@sydney.com",
    body: "est natus enim nihil est "
  },
  {
    postId: 1,
    id: 3,
    name: "odio adipisci",
    email: "Nikita@garfield.biz",
    body: "quia molestiae reprehenderit "
  }
];
