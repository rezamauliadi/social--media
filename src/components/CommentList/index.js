import React, { Component } from "react";
import PropTypes from "prop-types";

import { Button, Comment, Label, Loader } from "semantic-ui-react";
import CommentItem from "src/components/CommentItem";
import CreateForm from "src/components/CreateForm";

import API from "src/api";

class CommentList extends Component {
  state = {
    comments: [],
    showComments: false,
    hasFetchComments: false,
    fetchingComments: false,
    commentButtonText: "Show Comments"
  };

  getComments = async () => {
    const { postId } = this.props;
    const comments = await API.retrieveCommentsByPost(postId);
    return comments;
  };

  pushNewComment = newComment => {
    const comments = [...this.state.comments];
    comments.push(newComment);
    this.setState({ comments });
  };

  updateComment = updatedComment => {
    const { comments } = this.state;
    const commentIndex = comments.findIndex(
      comment => comment.id === updatedComment.id
    );
    const newCommenttList = [...comments];
    newCommenttList.splice(commentIndex, 1, updatedComment);
    this.setState({ comments: newCommenttList });
  };

  removeComment = commentId => {
    const { comments } = this.state;
    const commentIndex = comments.findIndex(
      comment => comment.id === commentId
    );
    const newCommenttList = [...comments];
    newCommenttList.splice(commentIndex, 1);
    this.setState({ comments: newCommenttList });
  };

  toggleComments = async () => {
    const { comments, hasFetchComments, showComments } = this.state;

    if (!showComments) {
      if (comments.length || (!comments.length && hasFetchComments)) {
        this.setState({
          showComments: true,
          commentButtonText: "Hide Comments"
        });
      } else {
        this.setState({ fetchingComments: true });
        const comments = await this.getComments();
        this.setState({
          comments,
          commentButtonText: "Hide Comments",
          fetchingComments: false,
          hasFetchComments: true,
          showComments: true
        });
      }
    } else {
      this.setState({
        showComments: false,
        commentButtonText: "Show Comments"
      });
    }
  };

  commentButtonContent = () => {
    const { fetchingComments, commentButtonText } = this.state;

    if (fetchingComments) {
      return (
        <div>
          <Loader style={{ marginRight: "4px" }} size="mini" active inline />{" "}
          Loading...
        </div>
      );
    }
    return commentButtonText;
  };

  createCommentForm = () => {
    return (
      <div style={{ marginTop: "18px" }}>
        <CreateForm type="comments" onSubmit={this.pushNewComment} />
      </div>
    );
  };

  commentList = () => {
    const { comments } = this.state;
    const { postId } = this.props;
    return comments.map(comment => (
      <CommentItem
        comment={comment}
        key={`${postId}-${comment.id}`}
        onUpdateComment={this.updateComment}
        onDeleteComment={this.removeComment}
      />
    ));
  };

  renderComments = () => {
    const { comments, hasFetchComments, showComments } = this.state;

    if (comments.length && showComments) {
      return (
        <div>
          {this.commentList()}
          {this.createCommentForm()}
        </div>
      );
    } else if (!comments.length && hasFetchComments && showComments) {
      return (
        <div>
          <Label basic>No comment yet...</Label>
          {this.createCommentForm()}
        </div>
      );
    }
    return <div />;
  };

  render() {
    return (
      <div>
        <Button
          size="mini"
          style={{ marginLeft: "50px" }}
          onClick={() => this.toggleComments()}
        >
          {this.commentButtonContent()}
        </Button>
        <Comment.Group style={{ marginLeft: "60px" }}>
          {this.renderComments()}
        </Comment.Group>
      </div>
    );
  }
}

CommentList.propTypes = {
  postId: PropTypes.number
};

export default CommentList;
