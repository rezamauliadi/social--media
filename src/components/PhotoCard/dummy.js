const albums = [
  {
    userId: 1,
    id: 1,
    title: "quidem molestiae enim"
  },
  {
    userId: 2,
    id: 11,
    title: "quidem molestiae enim"
  }
];
const photos = [
  {
    albumId: 1,
    id: 1,
    title: "accusamus beatae ad facilis cum similique qui sunt",
    url: "https://via.placeholder.com/600/92c952",
    thumbnailUrl: "https://via.placeholder.com/150/92c952"
  },
  {
    albumId: 2,
    id: 51,
    title: "non sunt voluptatem placeat consequuntur rem incidunt",
    url: "https://via.placeholder.com/600/8e973b",
    thumbnailUrl: "https://via.placeholder.com/150/8e973b"
  }
];

export default { albums, photos };
