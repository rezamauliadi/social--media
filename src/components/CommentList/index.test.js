import React from "react";
import { shallow } from "enzyme";

import CommentList from "./index.js";
import dummy from "./dummy.js";
import API from "src/api";

describe("Component: CommentList", () => {
  const props = { postId: 1 };

  it("should match snapshot", () => {
    const component = shallow(<CommentList {...props} />);
    expect(component).toMatchSnapshot();
  });

  describe("button call", () => {
    it("should call submitItem when submit button clicked", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "toggleComments");
      component
        .find("Button")
        .at(0)
        .simulate("click");
      expect(instance.toggleComments).toHaveBeenCalled();
    });
  });

  describe("renderComments", () => {
    it("should match snapshot if has comments", () => {
      const component = shallow(<CommentList {...props} />);
      component.setState({ comments: dummy, showComments: true });

      expect(component).toMatchSnapshot();
    });

    it("should match snapshot if has no comments and has fetch comment", () => {
      const component = shallow(<CommentList {...props} />);
      component.setState({
        comments: [],
        hasFetchComments: true,
        showComments: true
      });

      expect(component).toMatchSnapshot();
    });
  });

  describe("toggleComments", () => {
    it("should set correct button text when has comments", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      component.setState({ comments: dummy, showComments: false });
      instance.toggleComments();

      expect(component.state("showComments")).toBe(true);
      expect(component.state("commentButtonText")).toBe("Hide Comments");
    });

    it("should set correct button text when has no comments and has fetch", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      component.setState({
        comments: dummy,
        showComments: false,
        hasFetchComments: true
      });
      instance.toggleComments();

      expect(component.state("showComments")).toBe(true);
      expect(component.state("commentButtonText")).toBe("Hide Comments");
    });

    it("should call getComments and set correct button text when has no comments", done => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "getComments").mockImplementation(() => dummy);
      component.setState({
        comments: [],
        showComments: false,
        hasFetchComments: false
      });
      instance.toggleComments();

      process.nextTick(() => {
        expect(instance.getComments).toHaveBeenCalled();
        expect(component.state("showComments")).toBe(true);
        expect(component.state("commentButtonText")).toBe("Hide Comments");
        done();
      });
    });

    it("should set correct button text when showComment is true", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      component.setState({ showComments: true });
      instance.toggleComments();

      expect(component.state("showComments")).toBe(false);
      expect(component.state("commentButtonText")).toBe("Show Comments");
    });
  });

  describe("removeComment", () => {
    it("should remove specific comment by id", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      const commentId = 2;
      const commentIndex = 1;
      const expected = [...dummy];
      expected.splice(commentIndex, 1);

      component.setState({ comments: dummy });
      instance.removeComment(commentId);

      expect(component.state("comments")).toEqual(expected);
    });
  });

  describe("updateComment", () => {
    const updatedComment = {
      postId: 1,
      id: 2,
      name: "updated",
      body: "updated",
      email: "email@email.com"
    };

    it("should update specific comment by id", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      const commentIndex = 1;
      const expected = [...dummy];
      expected.splice(commentIndex, 1, updatedComment);

      component.setState({ comments: dummy });
      instance.updateComment(updatedComment);

      expect(component.state("comments")).toEqual(expected);
    });
  });

  describe("pushNewComment", () => {
    const newComment = {
      postId: 1,
      id: 4,
      name: "new",
      body: "new",
      email: "email@email.com"
    };

    it("should add new comment", () => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      const expected = [...dummy];
      expected.push(newComment);

      component.setState({ comments: dummy });
      instance.pushNewComment(newComment);

      expect(component.state("comments")).toEqual(expected);
    });
  });

  describe("getComments", () => {
    it("should call retrieveCommentsByPost", done => {
      const component = shallow(<CommentList {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "retrieveCommentsByPost").mockImplementation(() => dummy);

      instance.getComments();
      process.nextTick(() => {
        expect(API.retrieveCommentsByPost).toHaveBeenCalled();
        done();
      });
    });
  });
});
