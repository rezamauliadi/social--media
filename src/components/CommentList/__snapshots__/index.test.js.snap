// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CommentList renderComments should match snapshot if has comments 1`] = `
<div>
  <Button
    as="button"
    onClick={[Function]}
    size="mini"
    style={
      Object {
        "marginLeft": "50px",
      }
    }
  >
    Show Comments
  </Button>
  <CommentGroup
    style={
      Object {
        "marginLeft": "60px",
      }
    }
  >
    <div>
      <CommentItem
        comment={
          Object {
            "body": "laudantium ",
            "email": "Eliseo@gardner.biz",
            "id": 1,
            "name": "id labore",
            "postId": 1,
          }
        }
        key="1-1"
        onDeleteComment={[Function]}
        onUpdateComment={[Function]}
      />
      <CommentItem
        comment={
          Object {
            "body": "est natus enim nihil est ",
            "email": "Jayne_Kuhic@sydney.com",
            "id": 2,
            "name": "quo vero ",
            "postId": 1,
          }
        }
        key="1-2"
        onDeleteComment={[Function]}
        onUpdateComment={[Function]}
      />
      <CommentItem
        comment={
          Object {
            "body": "quia molestiae reprehenderit ",
            "email": "Nikita@garfield.biz",
            "id": 3,
            "name": "odio adipisci",
            "postId": 1,
          }
        }
        key="1-3"
        onDeleteComment={[Function]}
        onUpdateComment={[Function]}
      />
      <div
        style={
          Object {
            "marginTop": "18px",
          }
        }
      >
        <CreateForm
          onSubmit={[Function]}
          type="comments"
        />
      </div>
    </div>
  </CommentGroup>
</div>
`;

exports[`Component: CommentList renderComments should match snapshot if has no comments and has fetch comment 1`] = `
<div>
  <Button
    as="button"
    onClick={[Function]}
    size="mini"
    style={
      Object {
        "marginLeft": "50px",
      }
    }
  >
    Show Comments
  </Button>
  <CommentGroup
    style={
      Object {
        "marginLeft": "60px",
      }
    }
  >
    <div>
      <Label
        basic={true}
      >
        No comment yet...
      </Label>
      <div
        style={
          Object {
            "marginTop": "18px",
          }
        }
      >
        <CreateForm
          onSubmit={[Function]}
          type="comments"
        />
      </div>
    </div>
  </CommentGroup>
</div>
`;

exports[`Component: CommentList should match snapshot 1`] = `
<div>
  <Button
    as="button"
    onClick={[Function]}
    size="mini"
    style={
      Object {
        "marginLeft": "50px",
      }
    }
  >
    Show Comments
  </Button>
  <CommentGroup
    style={
      Object {
        "marginLeft": "60px",
      }
    }
  >
    <div />
  </CommentGroup>
</div>
`;
